File: dotnet/src/EPostak/Resources/WebhooksResource.cs
using EPostak.Models;

namespace EPostak.Resources;

/// <summary>
/// Manage webhook subscriptions for push-based event delivery, or use the
/// <see cref="Queue"/> for pull-based polling. Webhooks notify your server
/// when documents are created, sent, received, or validated.
/// </summary>
public sealed class WebhooksResource
{
    private readonly HttpRequestor _http;

    /// <summary>
    /// Access the webhook pull queue for polling-based event consumption.
    /// Use this instead of push webhooks if you cannot expose a public endpoint.
    /// </summary>
    public WebhookQueueResource Queue { get; }

    internal WebhooksResource(HttpRequestor http)
    {
        _http = http;
        Queue = new WebhookQueueResource(http);
    }

    /// <summary>
    /// Create a new webhook subscription. The response includes a one-time <c>Secret</c>
    /// field containing the HMAC-SHA256 signing key -- store it securely, as it cannot
    /// be retrieved again.
    /// </summary>
    /// <param name="request">The webhook URL and list of event types to subscribe to.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The created webhook with its signing secret (only returned on creation).</returns>
    /// <example>
    /// <code>
    /// var webhook = await client.Webhooks.CreateAsync(new CreateWebhookRequest
    /// {
    ///     Url = "https://example.com/webhooks/epostak",
    ///     Events = new List&lt;string&gt; { WebhookEvents.DocumentReceived, WebhookEvents.DocumentSent }
    /// });
    /// // Store webhook.Secret securely -- it won't be returned again
    /// Console.WriteLine($"Webhook {webhook.Id} created, secret: {webhook.Secret}");
    /// </code>
    /// </example>
    public Task<WebhookDetail> CreateAsync(CreateWebhookRequest request, CancellationToken ct = default)
        => CreateAsync(request, idempotencyKey: null, ct);

    /// <summary>
    /// Create a new webhook subscription with an optional <c>Idempotency-Key</c>
    /// header so retried calls return the original webhook (and signing secret)
    /// instead of provisioning a duplicate subscription.
    /// </summary>
    /// <param name="request">The webhook URL and list of event types to subscribe to.</param>
    /// <param name="idempotencyKey">Optional idempotency key for safe retries. Null disables the header.</param>
    /// <param name="ct">Cancellation token.</param>
    public Task<WebhookDetail> CreateAsync(CreateWebhookRequest request, string? idempotencyKey, CancellationToken ct = default)
        => _http.RequestAsync<WebhookDetail>(HttpMethod.Post, "/webhooks", request, idempotencyKey, ct);

    /// <summary>
    /// List all webhook subscriptions for the current API key.
    /// </summary>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>All webhook subscriptions with their URL, events, and active status.</returns>
    /// <example>
    /// <code>
    /// var webhooks = await client.Webhooks.ListAsync();
    /// foreach (var wh in webhooks)
    ///     Console.WriteLine($"{wh.Id}: {wh.Url} (active: {wh.IsActive})");
    /// </code>
    /// </example>
    public async Task<List<Webhook>> ListAsync(CancellationToken ct = default)
    {
        var res = await _http.RequestAsync<WebhookListResponse>(HttpMethod.Get, "/webhooks", ct).ConfigureAwait(false);
        return res.Data;
    }

    /// <summary>
    /// Get a webhook subscription with its recent delivery history.
    /// Use this to debug delivery failures and inspect response status codes.
    /// </summary>
    /// <param name="id">The webhook subscription UUID.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Webhook details including recent deliveries with status and attempt counts.</returns>
    /// <example>
    /// <code>
    /// var wh = await client.Webhooks.GetAsync("wh_abc123");
    /// foreach (var delivery in wh.Deliveries)
    ///     Console.WriteLine($"  {delivery.Event}: {delivery.Status} (attempts: {delivery.Attempts})");
    /// </code>
    /// </example>
    public Task<WebhookWithDeliveries> GetAsync(string id, CancellationToken ct = default)
        => _http.RequestAsync<WebhookWithDeliveries>(HttpMethod.Get, $"/webhooks/{Uri.EscapeDataString(id)}", ct);

    /// <summary>
    /// Update a webhook subscription. Change the URL, subscribed events, or active status.
    /// Only provided fields are updated; null fields remain unchanged.
    /// </summary>
    /// <param name="id">The webhook subscription UUID.</param>
    /// <param name="request">Fields to update: URL, events, and/or active status.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The updated webhook subscription.</returns>
    /// <example>
    /// <code>
    /// var updated = await client.Webhooks.UpdateAsync("wh_abc123", new UpdateWebhookRequest
    /// {
    ///     IsActive = false  // Pause the webhook
    /// });
    /// </code>
    /// </example>
    public Task<Webhook> UpdateAsync(string id, UpdateWebhookRequest request, CancellationToken ct = default)
        => _http.RequestAsync<Webhook>(HttpMethod.Patch, $"/webhooks/{Uri.EscapeDataString(id)}", request, ct);

    /// <summary>
    /// Delete a webhook subscription permanently. Pending deliveries will be cancelled.
    /// </summary>
    /// <param name="id">The webhook subscription UUID to delete.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>A completed task on success.</returns>
    /// <example>
    /// <code>
    /// await client.Webhooks.DeleteAsync("wh_abc123");
    /// </code>
    /// </example>
    public Task DeleteAsync(string id, CancellationToken ct = default)
        => _http.RequestVoidAsync(HttpMethod.Delete, $"/webhooks/{Uri.EscapeDataString(id)}", ct);

    /// <summary>
    /// Send a test event to a webhook endpoint. Useful for verifying your
    /// webhook URL is reachable and responding correctly.
    /// </summary>
    /// <param name="id">The webhook subscription UUID to test.</param>
    /// <param name="webhookEvent">Optional event type to simulate (e.g. "document.created").</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Test result with success status, HTTP status code, and response time.</returns>
    /// <example>
    /// <code>
    /// var result = await client.Webhooks.TestAsync("wh_abc123");
    /// Console.WriteLine($"Success: {result.Success}, Time: {result.ResponseTime}ms");
    /// </code>
    /// </example>
    public Task<WebhookTestResponse> TestAsync(string id, string? webhookEvent = null, CancellationToken ct = default)
    {
        var qs = webhookEvent != null ? $"?event={Uri.EscapeDataString(webhookEvent)}" : "";
        var body = new Dictionary<string, string>();
        if (webhookEvent != null) body["event"] = webhookEvent;
        return _http.RequestAsync<WebhookTestResponse>(HttpMethod.Post, $"/webhooks/{Uri.EscapeDataString(id)}/test{qs}", body, ct);
    }

    /// <summary>
    /// Send a test event to a webhook endpoint using a typed <see cref="WebhookTestParams"/>.
    /// The event type is sent as a <c>?event=</c> query parameter (server-side gives it
    /// precedence over the body field).
    /// </summary>
    /// <param name="id">The webhook subscription UUID to test.</param>
    /// <param name="params">Test parameters including the optional event type enum value.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Test result with success status, HTTP status code, and response time.</returns>
    /// <example>
    /// <code>
    /// var result = await client.Webhooks.TestAsync("wh_abc123", new WebhookTestParams
    /// {
    ///     Event = WebhookEvent.DocumentDelivered
    /// });
    /// Console.WriteLine($"Success: {result.Success}, Time: {result.ResponseTime}ms");
    /// </code>
    /// </example>
    public Task<WebhookTestResponse> TestAsync(string id, WebhookTestParams? @params, CancellationToken ct = default)
    {
        var wireEvent = @params?.Event != null ? WebhookEventToString(@params.Event.Value) : null;
        return TestAsync(id, wireEvent, ct);
    }

    /// <summary>Convert a <see cref="WebhookEvent"/> enum value to its wire string (e.g. <c>"document.delivered"</c>).</summary>
    private static string WebhookEventToString(WebhookEvent e) => e switch
    {
        WebhookEvent.DocumentCreated => WebhookEvents.DocumentCreated,
        WebhookEvent.DocumentSent => WebhookEvents.DocumentSent,
        WebhookEvent.DocumentReceived => WebhookEvents.DocumentReceived,
        WebhookEvent.DocumentValidated => WebhookEvents.DocumentValidated,
        WebhookEvent.DocumentDelivered => WebhookEvents.DocumentDelivered,
        WebhookEvent.DocumentRejected => WebhookEvents.DocumentRejected,
        WebhookEvent.DocumentResponseReceived => WebhookEvents.DocumentResponseReceived,
        _ => throw new ArgumentOutOfRangeException(nameof(e), e, "Unknown WebhookEvent value")
    };

    /// <summary>
    /// Get paginated delivery history for a webhook. Use this to inspect
    /// individual delivery attempts, filter by status, and debug failures.
    /// </summary>
    /// <param name="id">The webhook subscription UUID.</param>
    /// <param name="parameters">Optional pagination and filter parameters.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Paginated list of delivery records with total count.</returns>
    /// <example>
    /// <code>
    /// var result = await client.Webhooks.DeliveriesAsync("wh_abc123", new WebhookDeliveriesParams
    /// {
    ///     Status = "FAILED",
    ///     Limit = 50
    /// });
    /// foreach (var d in result.Deliveries)
    ///     Console.WriteLine($"{d.Event}: {d.Status} ({d.Attempts} attempts)");
    /// </code>
    /// </example>
    public Task<WebhookDeliveriesResponse> DeliveriesAsync(string id, WebhookDeliveriesParams? parameters = null, CancellationToken ct = default)
    {
        var query = new List<string>();
        if (parameters?.Limit != null) query.Add($"limit={parameters.Limit}");
        if (parameters?.Offset != null) query.Add($"offset={parameters.Offset}");
        if (parameters?.Status != null) query.Add($"status={Uri.EscapeDataString(parameters.Status)}");
        if (parameters?.Event != null) query.Add($"event={Uri.EscapeDataString(parameters.Event)}");
        var qs = query.Count > 0 ? "?" + string.Join("&", query) : "";
        return _http.RequestAsync<WebhookDeliveriesResponse>(HttpMethod.Get, $"/webhooks/{Uri.EscapeDataString(id)}/deliveries{qs}", ct);
    }

    /// <summary>
    /// Rotate a webhook's HMAC-SHA256 signing secret. Issues a fresh secret
    /// and invalidates the previous one immediately. The new secret is
    /// returned ONCE — store it right away; there is no way to retrieve it
    /// later. Any in-flight deliveries signed with the old secret will stop
    /// verifying on the receiving side. Non-destructive alternative to
    /// delete+recreate when a secret leaks.
    /// </summary>
    /// <param name="id">Webhook UUID whose secret to rotate.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The new signing secret (only shown once) and a confirmation message.</returns>
    /// <example>
    /// <code>
    /// var res = await client.Webhooks.RotateSecretAsync("wh_abc123");
    /// secretsManager.Save("epostak_webhook_secret", res.Secret);
    /// </code>
    /// </example>
    public Task<WebhookRotateSecretResponse> RotateSecretAsync(string id, CancellationToken ct = default)
    {
        return _http.RequestAsync<WebhookRotateSecretResponse>(
            HttpMethod.Post,
            $"/webhooks/{Uri.EscapeDataString(id)}/rotate-secret",
            ct);
    }

}

File: java/src/main/java/sk/epostak/sdk/resources/WebhooksResource.java
package sk.epostak.sdk.resources;

import sk.epostak.sdk.HttpClient;
import sk.epostak.sdk.models.Webhook;
import sk.epostak.sdk.models.WebhookDeliveriesResponse;
import sk.epostak.sdk.models.WebhookDetail;
import sk.epostak.sdk.models.WebhookRotateSecretResponse;
import sk.epostak.sdk.models.WebhookTestResponse;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.StringJoiner;

/**
 * Manage webhook subscriptions for push-based event delivery.
 * <p>
 * Supports creating, listing, updating, and deleting webhook endpoints.
 * For polling-based consumption, use the {@link #queue()} sub-resource instead.
 * <p>
 * Access via {@code client.webhooks()}.
 *
 * <pre>{@code
 * // Register a webhook for document events
 * WebhookDetail hook = client.webhooks().create(
 *     "https://example.com/webhook",
 *     List.of("document.received", "document.delivered"));
 * System.out.println("Secret: " + hook.secret()); // Save this for signature verification
 * }</pre>
 */
public final class WebhooksResource {

    private final HttpClient http;
    private final WebhookQueueResource queue;

    /**
     * Creates a new webhooks resource.
     *
     * @param http the HTTP client used for API communication
     */
    public WebhooksResource(HttpClient http) {
        this.http = http;
        this.queue = new WebhookQueueResource(http);
    }

    /**
     * Access the webhook pull queue for polling-based event consumption.
     *
     * @return the queue sub-resource
     */
    public WebhookQueueResource queue() {
        return queue;
    }

    /**
     * Register a new webhook endpoint. The response includes an HMAC-SHA256
     * signing secret that should be stored securely for verifying webhook payloads.
     *
     * @param url    the webhook endpoint URL (must be HTTPS)
     * @param events list of event types to subscribe to (e.g. {@code "document.received"}),
     *               or {@code null} to subscribe to all events
     * @return the created webhook detail including the signing secret
     * @throws sk.epostak.sdk.EPostakException if the URL is invalid or the request fails
     */
    public WebhookDetail create(String url, List<String> events) {
        return create(url, events, null);
    }

    /**
     * Register a new webhook endpoint with an explicit {@code Idempotency-Key}
     * header. Replaying the same key while the original request is still in
     * flight returns HTTP 409 ({@code idempotency_conflict}).
     *
     * @param url            the webhook endpoint URL (must be HTTPS)
     * @param events         list of event types to subscribe to, or {@code null}
     * @param idempotencyKey opaque idempotency key (1-255 chars), or {@code null} to skip
     * @return the created webhook detail including the signing secret
     * @throws sk.epostak.sdk.EPostakException if the URL is invalid or the request fails
     */
    public WebhookDetail create(String url, List<String> events, String idempotencyKey) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("url", url);
        if (events != null) body.put("events", events);
        if (idempotencyKey != null) {
            return http.postIdempotent("/webhooks", body, WebhookDetail.class, idempotencyKey);
        }
        return http.post("/webhooks", body, WebhookDetail.class);
    }

    /**
     * Register a new webhook endpoint subscribing to all event types.
     *
     * @param url the webhook endpoint URL (must be HTTPS)
     * @return the created webhook detail including the signing secret
     * @throws sk.epostak.sdk.EPostakException if the URL is invalid or the request fails
     */
    public WebhookDetail create(String url) {
        return create(url, null);
    }

    /**
     * List all registered webhook endpoints.
     *
     * @return list of webhooks, or an empty list if none
     * @throws sk.epostak.sdk.EPostakException if the request fails
     */
    public List<Webhook> list() {
        WebhookListWrapper wrapper = http.get("/webhooks", WebhookListWrapper.class);
        return wrapper != null ? wrapper.data() : List.of();
    }

    /**
     * Get webhook detail including delivery history for recent events.
     *
     * @param id the webhook UUID
     * @return the webhook detail with delivery attempts
     * @throws sk.epostak.sdk.EPostakException if the webhook is not found or the request fails
     */
    public WebhookDetail get(String id) {
        return http.get("/webhooks/" + HttpClient.encode(id), WebhookDetail.class);
    }

    /**
     * Update a webhook endpoint. Only provided (non-null) fields are changed.
     *
     * @param id       the webhook UUID
     * @param url      new endpoint URL, or {@code null} to keep current
     * @param events   new event type list, or {@code null} to keep current
     * @param isActive set to {@code false} to pause delivery, or {@code null} to keep current
     * @return the updated webhook
     * @throws sk.epostak.sdk.EPostakException if the webhook is not found or the request fails
     */
    public Webhook update(String id, String url, List<String> events, Boolean isActive) {
        Map<String, Object> body = new LinkedHashMap<>();
        if (url != null) body.put("url", url);
        if (events != null) body.put("events", events);
        if (isActive != null) body.put("isActive", isActive);
        return http.patch("/webhooks/" + HttpClient.encode(id), body, Webhook.class);
    }

    /**
     * Delete a webhook endpoint. This is irreversible. Returns HTTP 204 on success.
     *
     * @param id the webhook UUID
     * @throws sk.epostak.sdk.EPostakException if the webhook is not found or the request fails
     */
    public void delete(String id) {
        http.deleteVoid("/webhooks/" + HttpClient.encode(id));
    }

    /**
     * Send a test event to a webhook endpoint. Useful for verifying your
     * webhook URL is reachable and responding correctly.
     *
     * @param id    the webhook UUID to test
     * @param event the event type to simulate (e.g. {@code "document.created"}), or {@code null} for the server default
     * @return the test result with success status, HTTP status code, and response time
     * @throws sk.epostak.sdk.EPostakException if the webhook is not found or the request fails
     */
    public WebhookTestResponse test(String id, String event) {
        Map<String, Object> body = new LinkedHashMap<>();
        if (event != null) body.put("event", event);
        return http.post("/webhooks/" + HttpClient.encode(id) + "/test", body, WebhookTestResponse.class);
    }

    /**
     * Send a test event to a webhook endpoint using the server default event type.
     *
     * @param id the webhook UUID to test
     * @return the test result
     * @throws sk.epostak.sdk.EPostakException if the webhook is not found or the request fails
     */
    public WebhookTestResponse test(String id) {
        return test(id, null);
    }

    /**
     * Get paginated delivery history for a webhook.
     *
     * @param id     the webhook UUID
     * @param params optional query parameters: {@code limit} (1-100), {@code offset},
     *               {@code status} (UPPERCASE: {@code PENDING}, {@code SUCCESS},
     *               {@code FAILED}, {@code RETRYING}), {@code event}
     * @return paginated delivery records with total count
     * @throws sk.epostak.sdk.EPostakException if the webhook is not found or the request fails
     */
    public WebhookDeliveriesResponse deliveries(String id, Map<String, Object> params) {
        StringJoiner qj = new StringJoiner("&", "?", "");
        qj.setEmptyValue("");
        if (params != null) {
            for (var entry : params.entrySet()) {
                if (entry.getValue() != null) {
                    qj.add(HttpClient.encode(entry.getKey()) + "=" + HttpClient.encode(String.valueOf(entry.getValue())));
                }
            }
        }
        return http.get("/webhooks/" + HttpClient.encode(id) + "/deliveries" + qj, WebhookDeliveriesResponse.class);
    }

    /**
     * Get paginated delivery history for a webhook with default parameters.
     *
     * @param id the webhook UUID
     * @return paginated delivery records
     * @throws sk.epostak.sdk.EPostakException if the webhook is not found or the request fails
     */
    public WebhookDeliveriesResponse deliveries(String id) {
        return deliveries(id, null);
    }

    /**
     * Rotate a webhook's HMAC-SHA256 signing secret. Issues a fresh secret
     * and invalidates the previous one immediately. The returned {@code secret}
     * is shown ONCE — store it right away; there is no way to retrieve it
     * later. In-flight deliveries signed with the old secret will no longer
     * verify on the receiving side. Non-destructive alternative to
     * delete+recreate when a secret leaks.
     *
     * @param id webhook UUID whose secret to rotate
     * @return the new signing secret (only shown once) and a confirmation message
     * @throws sk.epostak.sdk.EPostakException if the webhook is not found or the request fails
     */
    public WebhookRotateSecretResponse rotateSecret(String id) {
        return http.post(
            "/webhooks/" + HttpClient.encode(id) + "/rotate-secret",
            null,
            WebhookRotateSecretResponse.class
        );
    }

    // -- internal wrappers ----------------------------------------------------

    private record WebhookListWrapper(List<Webhook> data) {}
}
